[PR] 1차 과제 - 김병한

OikuraSodachi opened the pull request:
ViewHolder layout 은 의도적으로 엉성한 상태로 남겨두었습니다

SQLite 가 익숙치 않은 상태에서 하루 안에 작성하느라 미흡한 부분이 많을 거라 생각됩니다.
해당 부분 ( MyDatabaseHelper ) 위주로 리뷰 부탁드립니다

//---------
여기부터 2차

BaseRecyclerAdapter 부분 집중적으로 봐주시면 감사하겠습니다

OikuraSodachi committed:
feat

레이아웃 생성
데이터베이스 추가

OikuraSodachi committed:
fix

content_recycler_holder.xml 레이아웃 조정

OikuraSodachi committed:
chore

불필요한 gradle dependency 제거

OikuraSodachi committed:
refactor

insertData,updateData 요구 파라미터를 SampleData 로 변경

OikuraSodachi committed:
chore

SampleRepository 에 주입되는 context 를 ApplicationContext 로 변경

sungje365 commented on app/src/main/java/devcon/map/data/MyDatabaseHelper.kt:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    fun insertData(data: SampleData): Long {
+        return writableDatabase.insert(TABLE_NAME, null, values(data))
+    }
+
+    fun readData(): List<SampleData> {
+        val dataList = mutableListOf<SampleData>()
+        val cursor: Cursor = readableDatabase.query(TABLE_NAME, null, null, null, null, null, null)
+
+        with(cursor) {
+            while (moveToNext()) {
+                val id = getLong(getColumnIndexOrThrow(COLUMN_ID))
+                val title = getString(getColumnIndexOrThrow(COLUMN_TITLE))
+
+                dataList.add(SampleData(id, title))
+            }
+        }
+
+        cursor.close()
+        return dataList
+    }
+
+    fun updateData(data: SampleData): Int {
+        return writableDatabase.update(
+            TABLE_NAME,
+            values(data),
+            "$COLUMN_ID = ?",
+            selectionArgs(data.id)
+        )
+    }
+
+    /** @param id id of data to delete **/
+    fun deleteData(id: Long): Int {
+        return writableDatabase.delete(TABLE_NAME, "$COLUMN_ID = ?", selectionArgs(id))
+    }
```
일반적으로 데이터베이스를 사용하기위한 Open 작업은 비용이 많이 필요한 작업으로 이야기 하는데요.
CURD 함수를 호출할 때 마다 데이터베이스를 Open하면 오버헤드가 많이 발생하지 않을까요?

OikuraSodachi replied:
> Create and/or open a database that will be used for reading and writing. The first time this is called, the database will be opened and [onCreate](https://developer.android.com/reference/kotlin/android/database/sqlite/SQLiteOpenHelper#onCreate(android.database.sqlite.SQLiteDatabase)), [onUpgrade](https://developer.android.com/reference/kotlin/android/database/sqlite/SQLiteOpenHelper#onUpgrade(android.database.sqlite.SQLiteDatabase,%20kotlin.Int,%20kotlin.Int)) and/or [onOpen](https://developer.android.com/reference/kotlin/android/database/sqlite/SQLiteOpenHelper#onOpen(android.database.sqlite.SQLiteDatabase)) will be called.
> 
> **_Once opened successfully, the database is cached, so you can call this method every time you need to write to the database_**. (Make sure to call [close](https://developer.android.com/reference/kotlin/android/database/sqlite/SQLiteOpenHelper#close()) when you no longer need the database.) Errors such as bad permissions or a full disk may cause this method to fail, but future attempts may succeed if the problem is fixed.

https://developer.android.com/reference/kotlin/android/database/sqlite/SQLiteOpenHelper#getwritabledatabase

최초 호출 이후 cache 되기 때문에 그 부분은 문제되지 않을 듯 합니다

sungje365 commented on app/src/main/java/devcon/map/activity/MainActivity.kt:
```diff
@@ -0,0 +1,43 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        viewModel = ViewModelProvider(
+            this,
+            MainViewModelFactory(SampleRepository(applicationContext))
+        ).get(MainViewModel::class.java)
```
`ViewModelProvider` 를 사용하여 `ViewModel` 에 종속 항목을 주입할 수 있군요!
제가 고민했던 부분이라 도움됬어요 👍

sungje365 reviewed (commented):
고생하셨습니다. 👍

OikuraSodachi committed:
chore

BaseRecyclerAdapter 적용
 - RecyclerView.Adapter 부분 로직은 마무리

OikuraSodachi committed:
feat

검색 결과 목록 표시 기능
검색 결과, 검색 기록 저장 기능
기록 삭제 / 추가 기능

OikuraSodachi committed:
feat

버튼 클릭으로 editText 내용 clear 기능 추가

OikuraSodachi committed:
chore

BaseRecyclerAdapter 의 dataSet 갱신 로직 최적화

OikuraSodachi committed:
chore

BaseRecyclerAdapter 의 요구 파라미터를 BaseRecyclerViewHolder 에서 RecyclerView.ViewHolder 로 완화함

OikuraSodachi committed:
chore

BaseRecyclerAdapter 의 타입을 BaseRecyclerViewHolder 에서 RecyclerView.ViewHolder 로 완화함

OikuraSodachi committed:
chore

BaseRecyclerAdapter 의 itemList 를 private 으로 변경

sungje365 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -13,7 +13,8 @@
         <activity
-            android:name="devcon.map.MainActivity"
+            android:name="devcon.map.activity.MainActivity"
+            android:configChanges="orientation|screenSize|screenLayout|keyboardHidden"
```
`android:configChanges` 속성은 어떤 이유로 추가되었나요?

OikuraSodachi replied:
제가 configChange 에 의한 Activity 재생성을 몹시 싫어합니다...

sungje365 commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,19 +1,50 @@
+            android:id="@+id/clearButton"
+            android:layout_width="wrap_content"
+            android:layout_height="match_parent"
+            android:src="@android:drawable/ic_input_delete" />
```
`Material TextInputLayout` 을 사용하고 계신다면 입력한 내용을 지우기 위하여 [Clear text](https://github.com/material-components/material-components-android/blob/master/docs/components/TextField.md#adding-a-trailing-icon-to-a-text-field) 기능을 사용하는 것도 고려할 수 있어요!

sungje365 commented on app/src/main/java/devcon/map/data/SampleData.kt:
`검색 기록`과 `장소`가 같은 모델을 사용하고 있는데요.
예를 들어 장소에 필드가 추가되게 된다면 `검색 기록` 에서 불필요한 필드도 포함하는 문제가 생길 것 같아요.
추가적으로 클래스의 이름에 의미를 부여해주면 좋을 것 같아요!

sungje365 reviewed (commented):
스터디때 언급되지 않은 부분 위주로 리뷰하였습니다.
고생하셨습니다. 👍

OikuraSodachi committed:
feat

검색 화면으로 이동/검색 화면에서 복귀 기능 추가

OikuraSodachi committed:
feat
        searchInputLayout 에  app:endIconMode="clear_text" 속성 추가

OikuraSodachi committed:
refactor

viewModel 을 by lazy 로 선언

OikuraSodachi committed:
refactor

KakaoMap sdk 관리를 위해 KakaoMapActivity 생성

OikuraSodachi committed:
refactor

MapView 생성

OikuraSodachi committed:
fix

일부 레이아웃의 background 가 투명하던 현상 fix

OikuraSodachi committed:
chore

키워드 검색 response 들어오는거 확인됨. 일단 백업용으로 commit

OikuraSodachi committed:
chore

키워드 검색 response 들어오는거 확인됨. 일단 백업용으로 commit

OikuraSodachi committed:
chore

키워드 검색 response 들어오는거 확인됨. 일단 백업용으로 commit

OikuraSodachi committed:
chore

키워드 검색 response 들어오는거 확인됨. 일단 백업용으로 commit

OikuraSodachi committed:
refactor

REST api key 전달

OikuraSodachi committed:
refactor

SampleRepository -> HistoryRepository rename
REST api key 전달

OikuraSodachi committed:
feat

15 개 이상의 키워드 검색 결과 구현 확인됨.

OikuraSodachi committed:
feat

15 개 이상의 키워드 검색 결과 구현 확인됨.

OikuraSodachi committed:
fix

텍스트 입력 과정에서 검색 결과가 엉키는 현상 해결

OikuraSodachi committed:
refactor

불필요한 값 제거

OikuraSodachi committed:
chore

주석 수정
